File: okhttplib/src/main/java/com/example/okhttplib/view/CountDownView.java
package com.example.okhttplib.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;

import java.util.Timer;
import java.util.TimerTask;

/**
 * Created by renhuan on
 * 倒计时控件 button
 */

@SuppressLint("DefaultLocale")
public class CountDownView extends android.support.v7.widget.AppCompatButton {

    private Timer timer;
    private int second = 60;
    @SuppressLint("HandlerLeak")
    private Handler handler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if (msg.what == 1) {
                setText(String.format("%d秒", second));
                second--;
                if (second == 0) {
                    timer.cancel();
                    timer = null;
                    setClickable(true);
                    setText("重新获取");
                    second = 60;
                }
            }
        }
    };

    public CountDownView(Context context) {
        this(context, null);
    }

    public CountDownView(Context context, AttributeSet attrs) {
        super(context, attrs);

    }

    /***
     * 开始计时
     */
    public void countDown() {
        setClickable(false);
        setText(String.format("%d", second));
        timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                handler.sendEmptyMessage(1);
            }
        }, 0, 1000);
    }
}

File: okhttplib/src/main/java/com/renhuan/okhttplib/view/CountDownView.kt
package com.renhuan.okhttplib.view

import android.content.Context
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatButton
import com.renhuan.okhttplib.R
import com.renhuan.okhttplib.utils.Renhuan
import com.renhuan.okhttplib.utils.postDelayed

/**
 * created by renhuan
 * time : 2020/5/9 13:07
 * describe : 倒计时控件
 */

class CountDownView(context: Context, attrs: AttributeSet? = null) : AppCompatButton(context, attrs) {

    /**
     * 倒计时最大值
     */
    private val maxTime = 60
    private var second = maxTime
    private var runnable: Runnable? = null

    /***
     * 开始计时
     */
    fun startCountDown() {
        isClickable = false
        runnable = Renhuan.getHandler().postDelayed(0, 1000) {
            text = second.toString()
            if (second-- == 0) {
                isClickable = true
                text = "重新获取"
                second = maxTime
                Renhuan.getHandler().removeCallbacks(runnable)
            }
        }
    }
}
